ui/src/App.tsx

```diff
@@ -0,0 +1,25 @@
+import React, { useEffect, useState } from "react";
+import { getJamState, JamState } from "./utils/jamState";
+import PageHeader from "./pages/PageHeader";
+import { AppRoutes } from "./AppRoutes";
+import { Toaster } from "react-hot-toast";
+
+export const App: React.FC = () => {
+  const [jamState, setJamState] = useState<JamState>(getJamState())
+
+  useEffect(() => {
+    // Periodically check state so that transitions between states happen automatically
+    setInterval(() => setJamState(getJamState()), 1000)
+  })
```

Comment:
This is actually going to start a new interval every time App re-renders - which will be every second. Timers and useEffect are tricky - although there's some research out there about how to avoid the footguns!

Reply:
I dang, I put some `console.log`'s around to catch rerenders and didn't catch anything 🤔 

I can always persist the timeout to a variable outside of the effect, but I feel like that's not quite what you meant?

Reply:
Aha, you know what you can do? Go ahead and throw a `[]` dependencies array on the `useEffect` (as the final argument). That way it'll only run the effect once - this is safe because it doesn't actually need anything from the function scope (it wouldn't be if the interval was a prop, for example)

I think it'd also be good to put a `clearInterval` as a cleanup function. Strictly speaking, I don't think `App` can get unmounted, but I prefer not to assume things :)

Reply:
So would that look like this?

```
useEffect(() => {
  // Periodically check state so that transitions between states happen automatically
  const jamStateCheck = setInterval(() => setJamState(getJamState()), 1000)
  return () => clearTimeout(jamStateCheck)
}, [])
```

Or do I need to have a `useRef` persist the interval outside of the effect?

Reply:
Nah, that works! The cleanup function captures the the interval in a closure.